futures-util/src/stream/stream/mod.rs

```diff
@@ -540,6 +544,37 @@ pub trait StreamExt: Stream {
         Flatten::new(self)
     }
 
+    /// Combinator similar to [`StreamExt::fold`] that holds internal state and produces a new stream.
+    ///
+    /// Accepts initial state and closure which will be applied to each element of the stream until provided closure
+    /// returns `None`. Once `None` is returned, stream will be terminated.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # futures::executor::block_on(async {
+    /// use futures::future;
+    /// use futures::stream::{self, StreamExt};
+    ///
+    /// let stream = stream::iter(1..=10);
+    ///
+    /// let stream = stream.scan(0, |state, x| {
+    ///     *state += x;
+    ///     future::ready(if *state < 10 { Some(x) } else { None })
+    /// });
+    ///
+    /// assert_eq!(vec![1, 2, 3], stream.collect::<Vec<_>>().await);
+    /// # });
+    /// ```
+    fn scan<S, B, Fut, F>(self, initial_state: S, f: F) -> Scan<Self, S, Fut, F>
+    where
+        F: FnMut(&mut S, Self::Item) -> Fut,
```

Comment:
There isn't a way to fix this at the moment, sadly, but this is subtly the wrong method signature for this function, because it doesn't allow `Fut` to hold onto `&mut S`. There's no way to fix this right now due to limitations of HRTBs / GATs, so I don't mind accepting this PR now, but note that this makes this API dramatically less useful than it could be otherwise.